index.js

```diff
@@ -26,59 +32,46 @@ class FirstChunkStream extends DuplexStream {
 		// Initialize the internal state
 		state.manager = createReadStreamBackpressureManager(this);
 
-		// Errors management
-		// We need to execute the callback or emit en error dependending on the fact
-		// the firstChunk is sent or not
-		state.errorHandler = error => {
-			processCallback(error, Buffer.concat(state.chunks, state.size), state.encoding);
-		};
-
-		this.on('error', state.errorHandler);
-
-		// Callback management
-		const processCallback = (error, buffer, encoding, done = () => {}) => {
-			// When doing sync writes + emitting an errror it can happen that
-			// Remove the error listener on the next tick if an error where fired
-			// to avoid unwanted error throwing
-			if (error) {
-				setImmediate(() => this.removeListener('error', state.errorHandler));
-			} else {
-				this.removeListener('error', state.errorHandler);
-			}
-
+		const processCallback = (buffer, encoding, done) => {
 			state.sent = true;
 
-			callback(error, buffer, encoding, (error, buffer, encoding) => {
-				if (error) {
-					setImmediate(() => this.emit('error', error));
-					return;
-				}
-
-				if (!buffer) {
+			(async () => {
+				let result;
+				try {
+					result = await callback(buffer, encoding);
+				} catch (error) {
+					setImmediate(() => {
+						this.emit('error', error);
+					});
 					done();
```

Comment:
Shouldn't `done()` be called in the `setImmediate` too?	You can actually just pass the error directly to `done()` instead of emitting the event. `done(error);`.

https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback_1

Reply:
`done` does not accept any arguments.

Reply:
Ok, I see now that the arguments are not passed around.